pootle/apps/pootle_fs/files.py

```diff
@@ -0,0 +1,229 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+import io
+import logging
+import os
+
+from translate.storage.factory import getclass
+
+from django.contrib.auth import get_user_model
+from django.utils.functional import cached_property
+
+from pootle_statistics.models import SubmissionTypes
+from pootle_store import models as store_models
+from pootle_store.models import FILE_WINS, POOTLE_WINS, Store
+
+
+logger = logging.getLogger(__name__)
+
+
+class FSFile(object):
+
+    def __init__(self, store_fs):
+        """
+        :param store_fs: ``FSStore`` object
+        """
+        from .models import StoreFS
+
+        if not isinstance(store_fs, StoreFS):
+            raise TypeError(
+                "pootle_fs.FSFile expects a StoreFS")
+        self.store_fs = store_fs
+        self.pootle_path = store_fs.pootle_path
+        self.path = store_fs.path
+
+    def __str__(self):
+        return "<%s: %s::%s>" % (
+            self.__class__.__name__, self.pootle_path, self.path)
+
+    def __hash__(self):
+        return hash(
+            "%s::%s::%s::%s"
+            % (self.path,
+               self.pootle_path,
+               self.store_fs.last_sync_hash,
+               self.store_fs.last_sync_revision))
+
+    def __eq__(self, other):
+        return hash(other) == hash(self)
+
+    @property
+    def exists(self):
+        return os.path.exists(self.file_path)
+
+    @property
+    def file_path(self):
+        return os.path.join(
+            self.store_fs.project.local_fs_path,
+            self.path.strip("/"))
+
+    @property
+    def fs_changed(self):
+        latest_hash = self.latest_hash
+        return (
+            latest_hash is not None
+            and (
+                latest_hash
+                != self.store_fs.last_sync_hash))
+
+    @property
+    def latest_hash(self):
```

Comment:
for the default localfs plugin that is what it uses, as anything else would be too costly, but this is supposed to be generic to any kind of fs, so with git, mercurial etc it will get the repo hash